File: tests/cpu/python/test_unary.py
"""Unary CPU binding tests — aligned with C++ test_unary.cpp.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python
    LD_LIBRARY_PATH=build/backends/cpu
    python -m pytest tests/cpu/python/test_unary.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)


class TestUnaryCPU:
    """Unary math operations — CPU backend."""

    def test_abs(self):
        a_np = np.array([-3, -1, 0, 2, 4], dtype=np.float64)
        result = ins.abs(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.abs(a_np), rtol=1e-10)

    def test_sqrt(self):
        a_np = np.array([1, 4, 9, 16], dtype=np.float64)
        result = ins.sqrt(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.sqrt(a_np), rtol=1e-10)

    def test_exp(self):
        a_np = np.array([0, 1, 2], dtype=np.float64)
        result = ins.exp(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.exp(a_np), rtol=1e-10)

    def test_log(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        result = ins.log(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.log(a_np), rtol=1e-10)

    def test_sin_cos_tan(self):
        a_np = np.array([0, 0.5, 1.0], dtype=np.float64)
        np.testing.assert_allclose(ins.sin(ins.from_numpy(a_np)).numpy(), np.sin(a_np), rtol=1e-10)
        np.testing.assert_allclose(ins.cos(ins.from_numpy(a_np)).numpy(), np.cos(a_np), rtol=1e-10)
        np.testing.assert_allclose(ins.tan(ins.from_numpy(a_np)).numpy(), np.tan(a_np), rtol=1e-10)

    def test_floor_ceil_round(self):
        a_np = np.array([1.2, 2.5, 3.7, -1.3], dtype=np.float64)
        np.testing.assert_allclose(
            ins.floor(ins.from_numpy(a_np)).numpy(), np.floor(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.ceil(ins.from_numpy(a_np)).numpy(), np.ceil(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.round(ins.from_numpy(a_np)).numpy(), np.round(a_np), rtol=1e-10
        )

    def test_sign(self):
        a_np = np.array([-3, -1, 0, 2, 4], dtype=np.float64)
        result = ins.sign(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.sign(a_np), rtol=1e-10)

    def test_negative(self):
        a_np = np.array([1, -2, 3], dtype=np.float64)
        result = ins.negative(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), -a_np, rtol=1e-10)

    def test_square(self):
        a_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.square(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), a_np**2, rtol=1e-10)

    def test_reciprocal(self):
        a_np = np.array([1, 2, 4, 5], dtype=np.float64)
        result = ins.reciprocal(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), 1.0 / a_np, rtol=1e-10)

    def test_log2_log10(self):
        a_np = np.array([1, 2, 8, 100], dtype=np.float64)
        np.testing.assert_allclose(
            ins.log2(ins.from_numpy(a_np)).numpy(), np.log2(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.log10(ins.from_numpy(a_np)).numpy(), np.log10(a_np), rtol=1e-10
        )

    def test_asin_acos_atan(self):
        a_np = np.array([-0.5, 0, 0.5], dtype=np.float64)
        np.testing.assert_allclose(
            ins.asin(ins.from_numpy(a_np)).numpy(), np.arcsin(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.acos(ins.from_numpy(a_np)).numpy(), np.arccos(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.atan(ins.from_numpy(a_np)).numpy(), np.arctan(a_np), rtol=1e-10
        )

    def test_sinh_cosh_tanh(self):
        a_np = np.array([-1, 0, 1], dtype=np.float64)
        np.testing.assert_allclose(
            ins.sinh(ins.from_numpy(a_np)).numpy(), np.sinh(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.cosh(ins.from_numpy(a_np)).numpy(), np.cosh(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.tanh(ins.from_numpy(a_np)).numpy(), np.tanh(a_np), rtol=1e-10
        )

    def test_deg2rad_rad2deg(self):
        a_np = np.array([0, 90, 180, 360], dtype=np.float64)
        np.testing.assert_allclose(
            ins.deg2rad(ins.from_numpy(a_np)).numpy(), np.deg2rad(a_np), rtol=1e-10
        )
        b_np = np.array([0, np.pi / 2, np.pi, 2 * np.pi], dtype=np.float64)
        np.testing.assert_allclose(
            ins.rad2deg(ins.from_numpy(b_np)).numpy(), np.rad2deg(b_np), rtol=1e-10
        )

    def test_exp2_expm1_log1p(self):
        a_np = np.array([0, 1, 2, 3], dtype=np.float64)
        np.testing.assert_allclose(
            ins.exp2(ins.from_numpy(a_np)).numpy(), np.exp2(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.expm1(ins.from_numpy(a_np)).numpy(), np.expm1(a_np), rtol=1e-10
        )
        np.testing.assert_allclose(
            ins.log1p(ins.from_numpy(a_np)).numpy(), np.log1p(a_np), rtol=1e-10
        )

    def test_cbrt(self):
        a_np = np.array([1, 8, 27], dtype=np.float64)
        result = ins.cbrt(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.cbrt(a_np), rtol=1e-10)

    def test_trunc(self):
        a_np = np.array([1.2, -2.7, 3.5], dtype=np.float64)
        result = ins.trunc(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.trunc(a_np), rtol=1e-10)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cpu/julia/test_unary.jl
# Unary CPU tests — aligned with C++ test_unary.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu julia tests/cpu/julia/test_unary.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Unary ===")

u = Insight.from_data([-3.0, -1.0, 0.0, 2.0, 4.0])

# abs
check("abs", Insight.abs_fn(u).ptr != C_NULL)

# sqrt
v = Insight.from_data([1.0, 4.0, 9.0, 16.0])
check("sqrt", Insight.sqrt_fn(v).ptr != C_NULL)

# exp
w = Insight.from_data([0.0, 1.0, 2.0])
check("exp", Insight.exp_fn(w).ptr != C_NULL)

# log
check("log", Insight.log_fn(w).ptr != C_NULL)

# sin
x = Insight.from_data([0.0, 0.5, 1.0])
check("sin", Insight.sin_fn(x).ptr != C_NULL)

# cos
check("cos", Insight.cos_fn(x).ptr != C_NULL)

# tan
check("tan", Insight.tan_fn(x).ptr != C_NULL)

# floor
y = Insight.from_data([1.2, 2.5, 3.7, -1.3])
check("floor", Insight.floor_fn(y).ptr != C_NULL)

# ceil
check("ceil", Insight.ceil_fn(y).ptr != C_NULL)

# round
check("round", Insight.round_fn(y).ptr != C_NULL)

# negative
check("negative", Insight.negative(u).ptr != C_NULL)

# exp2
z = Insight.linspace(0.1, 2.0, Int64(10), Insight.float32)
check("exp2", Insight.exp2_fn(z).ptr != C_NULL)

# expm1
check("expm1", Insight.expm1_fn(z).ptr != C_NULL)

# log1p
check("log1p", Insight.log1p_fn(z).ptr != C_NULL)

# cbrt
check("cbrt", Insight.cbrt_fn(z).ptr != C_NULL)

# reciprocal
check("reciprocal", Insight.reciprocal_fn(z).ptr != C_NULL)

# asinh
check("asinh", Insight.asinh_fn(z).ptr != C_NULL)

# acosh
w2 = Insight.from_data([1.5, 2.0, 3.0])
check("acosh", Insight.acosh_fn(w2).ptr != C_NULL)

# deg2rad
d = Insight.from_data([0.0, 90.0, 180.0, 360.0])
check("deg2rad", Insight.deg2rad_fn(d).ptr != C_NULL)

# rad2deg
r = Insight.from_data([0.0, 1.5707963, 3.1415926, 6.2831853])
check("rad2deg", Insight.rad2deg_fn(r).ptr != C_NULL)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end
